tests/draft2020-12/unevaluatedItems.json

```
@@ -485,5 +485,183 @@
+                "data": [ 1, 2, "foo" ],
+                "valid": false
+            }
+        ]
+    },
+    {
+        "description": "unevaluatedItems depends on nested contains",
+        "schema": {
+            "anyOf": [
+                {
+                    "prefixItems": [{"type": "integer"}]
+                },
+                {
+                    "contains": {"type": "string"}
+                }
+            ],
+            "unevaluatedItems": false
+        },
+        "tests": [
+            {
+                "description": "single integer is evaluated by prefixItems",
+                "data": [ 1 ],
+                "valid": true
+            },
+            {
+                "description": "single string is evaluated by contains",
+                "data": [ "foo" ],
+                "valid": true
+            },
+            {
+                "description": "second item is evaluated by contains",
+                "data": [ 1, "foo" ],
+                "valid": true
+            },
+            {
+                "description": "second item is not evaluated, contains fails",
+                "data": [ 1, 2 ],
+                "valid": false
+            },
+            {
+                "description": "second item is not evaluated, contains passes",
+                "data": [ 1, 2, "foo" ],
+                "valid": false
+            }
+        ]
+    },
+    {
+        "description": "unevaluatedItems depends on multiple nested contains",
+        "schema": {
+            "allOf": [
+                {
+                    "prefixItems": [true]
+                },
+                {
+                    "contains": {"multipleOf": 2}
+                }
+            ],
+            "anyOf": [
+                {
+                    "prefixItems": [true]
+                },
+                {
+                    "contains": {"multipleOf": 3}
+                }
+            ],
+            "unevaluatedItems": {
+                "multipleOf": 5
+            }
+        },
+        "tests": [
+            {
+                "description": "second item is evaluated by contains #1",
+                "data": [ 1, 2 ],
+                "valid": true
+            },
+            {
+                "description": "third item is evaluated by contains #2",
+                "data": [ 1, 2, 3 ],
+                "valid": true
+            },
+            {
+                "description": "items 3, 4 are evaluated by contains #2, #1",
+                "data": [ 1, 2, 3, 4 ],
+                "valid": true
+            },
+            {
+                "description": "item 5 not evaluated, passes unevaluatedItems",
+                "data": [ 1, 2, 3, 4, 5, 6 ],
+                "valid": true
+            },
+            {
+                "description": "item 5 not evaluated, fails unevaluatedItems",
+                "data": [ 1, 2, 3, 4, 7, 8 ],
+                "valid": false
+            }
+        ]
+    },
```

Most of the things covered in this test are already covered in other tests. The value of this test is testing that contains annotations from different schemas get merged and applied correctly. I suggest this simplification.

```suggestion
    {
        "description": "unevaluatedItems depends on multiple nested contains",
        "schema": {
            "allOf": [
                { "contains": { "multipleOf": 2 } }
                { "contains": { "multipleOf": 3 } }
            ],
            "unevaluatedItems": { "multipleOf": 5 }
        },
        "tests": [
            {
                "description": "5 not evaluated, passes unevaluatedItems",
                "data": [ 2, 3, 4, 5, 6 ],
                "valid": true
            },
            {
                "description": "7 not evaluated, fails unevaluatedItems",
                "data": [ 2, 3, 4, 7, 8 ],
                "valid": false
            }
        ]
    },
```

---

My comment above on `prefixItems` influencing the interaction between `contains` and `unevaluatedItems` applies here too. My implementation passed these tests without `prefixItems`, but failed when `prefixItems` was introduced.

---

Did your implementation fail this test for different reason than it failed the first test?

Since we are keeping `prefixItems` in the other test, I don't think we will need it here as well. The first test tests the interaction between `contains` and `prefixItems`. So, this test can focus on the interaction between `contains` and `contains`. It doesn't need to test `contains`+`prefixItems` again.

---

Fair enough - it failed for the same reason as in the first test. I'll accept your suggested change.

I think there is potentially room for a test case with multiple nesting of both `contains` and `prefixItems`, because the comparison of `contains` indices with those left over from `prefixItems` must consider that each set of indices can come from multiple evaluation paths. However my initial version of this test does not really focus on that problem - in particular because it uses same-length `prefixItems` arrays. If I come up with a clearer way of testing this, I'll post it as a separate PR.